filesystem/squashfs_superblock.ksy

```diff
@@ -0,0 +1,54 @@
+meta:
+  id: squashfs_superblock
+  title: SquashFS superblock
+  file-extension:
+    - squashfs
+    - snap
+  endian: le
+  license: CC-0
+doc-ref: https://github.com/AgentD/squashfs-tools-ng/blob/master/doc/format.txt
+seq:
+  - id: superblock
+    type: superblock
+types:
+  superblock:
+    seq:
+      - id: magic
+        contents: 'hsqs'
+      - id: inode_count
+        type: u4
+      - id: mod_time
+        type: u4
+      - id: block_size
+        type: u4
+      - id: frag_count
+        type: u4
+      - id: compressor
+        type: u2
+      - id: block_log
+        type: u2
+      - id: flags
+        type: u2
+      - id: id_count
+        type: u2
+      - id: version_major
+        contents: [ 4, 0 ]
```

Comment:
Shouldn't it be just parsed probably with later validation using `valid` or with dispatching between versions?

Reply:
The spec says it should not change.

version_major | u16 | The major version of the squashfs file format. Should always equal 4
-- | -- | --
version_minor | u16 | The minor version of the squashfs file format. Should always equal 0

What is the difference between using `valid` and just specifying the required value in `contents`?

Reply:
>What is the difference between using valid and just specifying the required value in contents?

when using `valid` they are parsed first as numbers and are available as such

Reply:
> Shouldn't it be just parsed probably with later validation using `valid` or with dispatching between versions?

> when using `valid` they are parsed first as numbers and are available as such

I fully support that. As @KOLANICH said, the only difference of using `type: u2` + `valid` is that the data type is the correct one. Right now, `version_major` is a byte array, and using `contents` you assert the actually parsed bytes to expected bytes. In fact, this snippet:

```yaml
seq:
  - id: version_major
    size: 2 # specifying `size` without `type` implies a byte array in KS
    valid:
      eq: '[4, 0]'
```

is an **exact** equivalent of this:

```yaml
seq:
  - id: version_major
    contents: [4, 0]
```

You can check that the generated code is the same in both cases (I recommend using the [devel Web IDE](https://ide.kaitai.io/devel/), as you don't have to worry about outdated versions).

But wait, isn't it actually wrong? The `version_major` should be an integer according to the spec (`u16` originally, which means `u2` in KS), not a byte array. If someone queries `struct.version_major`, he should get a number `4`, not `byte[] { 0x04, 0x00 }`.

So I think this better expresses the intent:

```yaml
seq:
  - id: version_major
    type: u2
    valid:
      eq: 4
```

On top of that, asserting the byte array `[4, 0]` is in my opinion more error-prone, because you have to know in which endianness the field is supposed to be parsed. Hardly anyone would notice if you messed this up and asserted `[0, 4]` instead by mistake.

And speaking of endianness, if by chance the format was using [calculated endianness](https://doc.kaitai.io/user_guide.html#calc-endian), you would have no option but do some sort of conditional branching (e.g. `{id: version_major_le, contents: [4, 0], if: is_little_endian}` + `{id: version_major_be, contents: [0, 4], if: not is_little_endian}`), and I think we can agree that working with an integer is a simpler and cleaner approach.

Reply:
I was not sure if `kaitai` is little endian by default. Now that the file is explicitly `le`, the check with value looks more readable, indeed.

Reply:
87b8a3665af9cb2f2bc81c098823d9b6faf597b9